ReadMe.md
# Application Note: Setup TLS connection with Trusted Objects' Secure Elements

## Prepare the project directory

    mkdir AN_TO_SE_tls_wifi && cd AN_TO_SE_tls_wifi

## Clone Trusted Object demo

    git clone git@github.com:TrustedObjects/TOAN_xSE_Setup_IoT_TLS .
    git checkout main

## Install MBED OS library and DISCO_L475VG_IOT01A_wifi library

    mbed deploy

## Check the directory tree

	.
	├── DISCO_L475VG_IOT01A_wifi/
	├── DISCO_L475VG_IOT01A_wifi.lib
	├── hal_nvm_mbed.cpp
	├── libTO/
	│      ├── doc
	│      ├── examples
	│      ├── include
	│      └── src
	├── main.cpp
	├── mbed_app.json
	├── mbed-os/
	│      ├── ...
	├── mbed-os.lib
	├── ReadMe.md
	├── sample_https_get.c
	├── sample_https.h
	├── to-protect/
	│      ├── Files4Flashing/
	│      ├── TOP_info.h
	│      └── TOP_technical_info.h
	└── TO_user_config.h

## Adapt the project to your configuration
### mbed_app.json

Edit ./mbed_app.json
You will have to change WiFi credentials, and remote server.

	{
		"config": {
			"wifi-ssid": {
				"help": "WiFi SSID",
				"value": "\"to be updated with your Wifi SSID\""
			},
			"wifi-password": {
				"help": "WiFi Password",
				"value": "\"to be updated with your Wifi password\""
			},
			"server-hostname": {
				"help": "Remote Server Host Name",
				"value": "\"localhost\""
			},
			"server-go-port": {
				"help": "CA Server TCP Port",
				"value": "4443"
			}
		}
	}

## Optional: Select your Secure Element (TO136 or to-protect)

Open POC_settings.h and #define POC_SECURE_ELEMENT_USED accordingly to your choice:
	- POC_SE_TO136, if you want to test with TO136
	- POC_SE_TOPROTECT (default), if you want to test with to-protect

If you select TO136, don't forget to mount the board Arduino_TO136
onto the Arduino R3 connector of the board DISCO_L475VG_IOT01A.

## Flash the SSE (to-protect version only)

For TO136, skip this step, as TO136 comes ready and personalized by AVNET.

For to-protect, you need first to flash the SSE (Software Secure Element) and
its secure storage

The script to-protect/Files4Flashing/Flash-TOProtect-and-Storage.sh is provided
It is based on ST MicroElectronics tool named "STM32_Programmer_CLI"

	cd to-protect/Files4Flashing
	./Flash-TOProtect-and-Storage.sh

It will flash these 2 files:

	- TO-Protect-eval.cortexm3.bin, the SSE
	- SecureStorage_1.bin, the personalization data

## Generate and flash the firmware

Generate the firmware

    mbed compile -m DISCO_L475VG_IOT01A -t GCC_ARM -DTO_USER_CONFIG

Flash the firmware

    mbed compile ... -f

or copy the generated .bin file to the virtual disk associated to the MBED Board

## Start the openssl server

	Open a shell and launch the following command (there is a tls_server.sh
	in the "server" subdirectory) :

		openssl.exe s_server -WWW -accept 4443 -cert CertsTrusted\ServerCertificate-01011E0002000001.pem -key CertsTrusted\ServerKeyPair-01011E0002000001.pem -Verify 10 -CAfile CertsTrusted\Trusted-Objects-RoussetProduction-Generic_Samples_AWS-CA-Certificate-0101100002000000.pem  -state -debug -msg

	The following message will be displayed by openssl.
	The stest server is ready to accept incoming connections:

		verify depth is 10, must return a certificate
		Using default temp DH parameters
		ACCEPT

## Test the firmware

	Open a terminal after flashing the board
	Terminal configuration MUST be 115200,n,8,1

	Example with mbed-cli:

		mbed sterm --baudrate 115200

	The firmware will do the following operations

		- Initialize Secure Element, and display SECURE Element version
		- Initialize and connect Wifi
		- Connect in TLS to remote server
		- Send HTTP GET request an get `Hello world !` answer

## UART output Log

	Press the reset (black) button on the STM32L475 board
	The following kind of log will be displayed:

	[mbed] Detecting connected targets/boards to your system...
	[mbed] Opening serial terminal to "DISCO_L475VG_IOT01A"
	--- Terminal on /dev/ttyACM0 - 115200,8,N,1 ---
	[t=0.000006]
	[t=0.001298] ************************************
	[t=0.005635] * Trusted Objects TLS Demonstrator *
	[t=0.009973] ************************************
	[t=0.014313] ********** Initialize Secure Element ***************************
	[t=0.341567] POC_display_se_infos: SE version: 2.4.6
	[t=0.346128] ********** Initialize and connect Wifi *************************
	[t=0.933548] WIFI Module Initialized.
	[t=0.982506] es-wifi module MAC Address : C4:7F:51:8D:17:86
	[t=0.987735] connecting to : Freebox-4F25F0
	[t=6.097588] es-wifi module connected
	[t=6.100841] es-wifi module got IP Address : 192.168.1.174
	[t=6.105985] ********** Connect in TLS to remote server *********************
	[t=6.112752] Converting Server remote address from [192.168.1.9]
	[t=6.163503] Trying to connect to Server: 192.168.1.9:4443 ...
	[t=6.446495] TCP Connection opened successfully.
	[t=6.664601] TOSE_helper_tls_init_session: Using session 0
	[t=6.670454] POC_connect_tls_server: HostName is only an IP address (not DNS), so do NOT activate SNI extension
	[t=6.680208] POC_connect_tls_server: using certificate slot 0
	[t=6.685566] TOSE_helper_tls_do_handshake_step: *** Flight 1 ***
	[t=6.691209] TOSE_helper_tls_do_handshake_step: SNI len : 0
	[t=6.697057] TOSE_helper_tls_do_handshake_step: ==> ClientHello
	[t=6.702582] TOSE_helper_tls_do_handshake_step: *** Flight 4 ***
	[t=7.119506] TOSE_helper_tls_do_handshake_step: <== ServerHello
	[t=7.125611] TOSE_helper_tls_do_handshake_step: Detected cipher suite: c023
	[t=7.132187] TOSE_helper_tls_do_handshake_step: <== Certificate
	[t=7.654507] TOSE_helper_tls_do_handshake_step: <== ServerKeyExchange
	[t=7.967813] TOSE_helper_tls_do_handshake_step: Client authentication not requested
	[t=7.975162] TOSE_helper_tls_do_handshake_step: <== ServerHelloDone
	[t=7.981481] TOSE_helper_tls_do_handshake_step: *** Flight 5 ***
	[t=8.272055] TOSE_helper_tls_do_handshake_step: ==> ClientKeyExchange
	[t=8.420970] TOSE_helper_tls_do_handshake_step: ==> ChangeCipherSpec
	[t=8.786955] TOSE_helper_tls_do_handshake_step: ==> Finished
	[t=8.792217] TOSE_helper_tls_do_handshake_step: *** Flight 6 ***
	[t=9.155507] TOSE_helper_tls_do_handshake_step: <== ChangeCipherSpec
	[t=9.165308] TOSE_helper_tls_do_handshake_step: <== Finished
	[t=9.172577] POC_connect_tls_server: TO_helper_tls_do_handshake OK
	[t=9.178341] ********** Send HTTP GET request an get `Hello world !` answer *
	[t=9.185134] POC_http_post: HTTP request: "
	GET /index.html HTTP/1.1


	"

	[t=9.556926] POC_http_post: response : "
	HTTP/1.0 200 ok
	Content-type: text/html

	<html>
	<header><title>Hello</title></header>
	<body>
	Hello, world.
	</body>
	</html>
	"

	[t=9.615491] es-wifi disconnected
	[t=9.853851] Terminated ! Press RESET (black) button to restart

## openssl output Log

	Using only informational "-state" option on the command line, you get
	this output in the server terminal:

	openssl.exe s_server -WWW -accept 4443 -cert CertsTrusted\ServerCertificate-01011E0002000001.pem -key CertsTrusted\ServerKeyPair-01011E0002000001.pem -Verify 10 -CAfile CertsTrusted\Trusted-Objects-RoussetProduction-Generic_Samples_AWS-CA-Certificate-0101100002000000.pem  -state
	ACCEPT
	SSL_accept:before SSL initialization
	SSL_accept:before SSL initialization
	SSL_accept:SSLv3/TLS read client hello
	SSL_accept:SSLv3/TLS write server hello
	SSL_accept:SSLv3/TLS write certificate
	SSL_accept:SSLv3/TLS write key exchange
	SSL_accept:SSLv3/TLS write server done
	SSL_accept:SSLv3/TLS write server done
	SSL_accept:SSLv3/TLS read client key exchange
	SSL_accept:SSLv3/TLS read change cipher spec
	SSL_accept:SSLv3/TLS read finished
	SSL_accept:SSLv3/TLS write change cipher spec
	SSL_accept:SSLv3/TLS write finished
	FILE:index.html
	ACCEPT

## openssl output Log (Verbose)

	Adding option "-debug -msg" on the command line, you get this verbose output:

	openssl.exe s_server -WWW -accept 4443 -cert CertsTrusted\ServerCertificate-01011E0002000001.pem -key CertsTrusted\ServerKeyPair-01011E0002000001.pem -Verify 10 -CAfile CertsTrusted\Trusted-Objects-RoussetProduction-Generic_Samples_AWS-CA-Certificate-0101100002000000.pem  -state -debug -msg
	verify depth is 10, must return a certificate
	Using default temp DH parameters
	ACCEPT
	SSL_accept:before SSL initialization
	read from 0x1e109f36010 [0x1e109f4a293] (5 bytes => 5 (0x5))
	0000 - 16 03 03 00 4b                                    ....K
	<<< ??? [length 0005]
	16 03 03 00 4b
	read from 0x1e109f36010 [0x1e109f4a298] (75 bytes => 75 (0x4B))
	0000 - 01 00 00 47 03 03 00 00-00 00 be a9 e2 42 73 df   ...G.........Bs.
	0010 - 02 89 61 15 1a ac a8 4c-32 54 ad 7e 92 20 b1 ad   ..a....L2T.~. ..
	0020 - f3 f9 e9 e2 d0 2a 00 00-04 00 ae c0 23 01 00 00   .....*......#...
	0030 - 1a 00 0d 00 04 00 02 04-03 00 0a 00 04 00 02 00   ................
	0040 - 17 00 0b 00 02 01 00 00-04 00 00                  ...........
	SSL_accept:before SSL initialization
	<<< TLS 1.2Handshake [length 004b], ClientHello
	01 00 00 47 03 03 00 00 00 00 be a9 e2 42 73 df
	02 89 61 15 1a ac a8 4c 32 54 ad 7e 92 20 b1 ad
	f3 f9 e9 e2 d0 2a 00 00 04 00 ae c0 23 01 00 00
	1a 00 0d 00 04 00 02 04 03 00 0a 00 04 00 02 00
	17 00 0b 00 02 01 00 00 04 00 00
	SSL_accept:SSLv3/TLS read client hello
	>>> ??? [length 0005]
	16 03 03 00 54
	>>> TLS 1.2Handshake [length 0054], ServerHello
	02 00 00 50 03 03 4b b6 6a a3 c6 fc ff 8d 57 b8
	92 1b 59 4e 7c 20 10 45 f9 8f 07 3f 0c 77 08 99
	a3 92 3f 2f b8 36 20 86 29 bd 5d 62 a1 a6 0c 66
	e4 0e 60 cf 04 c8 9a 95 70 6e 0f a7 2e 84 99 53
	77 d4 cb f8 e5 c5 c0 c0 23 00 00 08 00 0b 00 04
	03 00 01 02
	SSL_accept:SSLv3/TLS write server hello
	>>> ??? [length 0005]
	16 03 03 04 27
	>>> TLS 1.2Handshake [length 0427], Certificate
	0b 00 04 23 00 04 20 00 01 e9 30 82 01 e5 30 82
	01 8b a0 03 02 01 02 02 08 01 01 1e 00 02 00 00
	01 30 0a 06 08 2a 86 48 ce 3d 04 03 02 30 7f 31
	1c 30 1a 06 03 55 04 07 0c 13 47 65 6e 65 72 69
	63 5f 53 61 6d 70 6c 65 73 5f 41 57 53 31 18 30
	16 06 03 55 04 0a 0c 0f 54 72 75 73 74 65 64 2d
	4f 62 6a 65 63 74 73 31 1a 30 18 06 03 55 04 0b
	0c 11 52 6f 75 73 73 65 74 50 72 6f 64 75 63 74
	69 6f 6e 31 29 30 27 06 03 55 04 03 0c 20 54 72
	75 73 74 65 64 2d 4f 62 6a 65 63 74 73 20 30 31
	30 31 31 30 30 30 30 32 30 30 30 30 30 30 30 22
	18 0f 32 30 31 34 30 31 32 31 30 30 30 30 30 30
	5a 18 0f 33 30 30 30 31 32 32 39 32 33 35 39 35
	39 5a 30 2b 31 29 30 27 06 03 55 04 03 0c 20 54
	72 75 73 74 65 64 2d 4f 62 6a 65 63 74 73 20 30
	31 30 31 31 45 30 30 30 32 30 30 30 30 30 31 30
	59 30 13 06 07 2a 86 48 ce 3d 02 01 06 08 2a 86
	48 ce 3d 03 01 07 03 42 00 04 93 ae c7 32 dc bf
	2f 49 be a9 0b 4d 24 a6 d5 4c 56 7e c3 3d 21 a2
	a7 4e 04 af 46 b5 1c d7 b1 12 01 16 8e 1c 67 dc
	c1 be be 8e 19 a8 5f e9 ff f2 d3 e2 fa ac a3 64
	b1 e0 da cc 71 5a dc 5d f3 d2 a3 41 30 3f 30 0e
	06 03 55 1d 0f 01 01 ff 04 04 03 02 03 c8 30 0c
	06 03 55 1d 13 01 01 ff 04 02 30 00 30 1f 06 03
	55 1d 23 04 18 30 16 80 14 7b b5 84 fa ee 59 71
	7e 1c 05 47 3f 1e a5 d4 e6 04 f2 87 0e 30 0a 06
	08 2a 86 48 ce 3d 04 03 02 03 48 00 30 45 02 21
	00 b3 80 ec 82 b4 ba 2e 6b bd 09 79 5d dc 1a 52
	78 17 8d 8c b1 15 98 53 17 34 70 ef 7e be 49 63
	00 02 20 68 ea 40 b4 57 71 d9 03 7d 68 c0 06 c7
	ab 51 73 7d b1 f0 01 59 76 c7 6e 9b 3c 22 25 38
	f4 9d 97 00 02 31 30 82 02 2d 30 82 01 d3 a0 03
	02 01 02 02 08 01 01 10 00 02 00 00 00 30 0a 06
	08 2a 86 48 ce 3d 04 03 02 30 61 31 18 30 16 06
	03 55 04 0a 0c 0f 54 72 75 73 74 65 64 2d 4f 62
	6a 65 63 74 73 31 1a 30 18 06 03 55 04 0b 0c 11
	52 6f 75 73 73 65 74 50 72 6f 64 75 63 74 69 6f
	6e 31 29 30 27 06 03 55 04 03 0c 20 54 72 75 73
	74 65 64 2d 4f 62 6a 65 63 74 73 20 30 30 30 31
	31 30 30 30 30 30 30 30 30 30 30 30 30 22 18 0f
	32 30 31 34 30 31 32 31 30 30 30 30 30 30 5a 18
	0f 33 30 30 30 31 32 32 39 32 33 35 39 35 39 5a
	30 7f 31 1c 30 1a 06 03 55 04 07 0c 13 47 65 6e
	65 72 69 63 5f 53 61 6d 70 6c 65 73 5f 41 57 53
	31 18 30 16 06 03 55 04 0a 0c 0f 54 72 75 73 74

libTO/ChangeLog
7.0.7 - 2023-07-04
- Removed a distribution constraint to to-protect-common

7.0.6 - 2023-06-26
- Changed the priority of a log message in the load/store in the RAM driver

7.0.5 - 2023-06-20
- Fix a bug related to Driver-SSE selection

7.0.4 - 2023-06-20
- Fix the build system (use SSE-SERIAL only if needed)

7.0.3 - 2023-06-13
- Re-add the secure storage constraint for building the distribition package correctly

7.0.2 - 2023-06-07
- Change ECIES parameters (merge encryption and mac)

7.0.1 - 2023-06-05
- Removing TO_legacy.h

7.0.0 - 2023-06-01
- New API with output length introduced

6.3.10 - 2023-02-24
- Using the newly defined X509_MAXSIZE

6.3.9 - 2023-01-27
- Better default driver configuration
- Helpers: handling TODRV_HSE_DRIVER_DISABLE

6.3.8 - 2023-01-27
- Better default driver configuration

6.3.7 - 2023-01-23
- Update for new documentation generator version"

6.3.6 - 2023-01-03
- Doc: reformulation after review on TO136 reset while upgrading

6.3.4 - 2023-01-02
- Doc: broadcast loader PN: TOSF-LD1 -> TOSB-BC1
- Doc: status PIO and power optimisation updates
- Doc: add details about TOSE_init() and TOSE_fini() timings

6.3.3 - 2022-11-30
- Updated Makefile

6.3.1 - 2022-11-30
- Updated link constraints

6.3.1 - 2022-11-29
- Updated constraints

6.3.0 - 2022-11-29
- Added a supplemental version of the library enabling a "no-init" x86 version

6.2.13 - 2022-11-10
- Added stack measurement

6.2.12 - 2022-10-28
- Update TO-Protect HAL NVM examples to use new load/store API
- Examples files cleanup

6.2.11 - 2022-10-10
- Python utils: fixed log level setting API after latest libTO changes
- TOsh is initializing log level to -1 by default
- Created TOSH_LIBTO_LOG_LEVEL env to be able to configure TOsh behavior before
call

6.2.10 - 2022-10-10
- Updated selftests

6.2.9 - 2022-09-20
- Removed a useless parameter in selftests

6.2.8 - 2022-09-20
- Fixed constraints in configure.ac
- Added comments in selftest example

6.2.7 - 2022-08-26
- secure_storage_fd.c example file removed
- fix agressive find&replace done in 93d9c2a2

6.2.6 - 2022-08-17
- Renamed a constant to DTLS_xxx

6.2.5 - 2022-08-12
- Warnings fix

6.2.4 - 2022-07-18
- Just because a tiny mistake has been made, and we hate deleting tags

6.2.2 - 2022-07-06
- Doc: removed enums and defines lists spacing because it is no more required
with Sphinx  5.0.2

6.2.1 - 2022-06-29
- Removed warnings about logs

6.2.0 - 2022-06-17
- Renamed all delivered files to be as TO_xxx

6.1.14 - 2022-06-16
- Change some prototypes for better clarity

6.1.13 - 2022-06-08
- Include TO_cfg.h from TOSE_cfg.h
- TO_user_config.h is now included by TO_cfg.h

6.1.9 - 2022-04-20
- Update the documentation to reflect TOP_info.h and TOP_technical_info.h for
TO-Protect
- Enable helper_cert when not using HSE

6.1.7 - 2022-04-06
- CAPI: adjust CAPI max update size in link with the driver selected
- CAPI feature shall not depend of driver HSE or SSE
- TLS helper: split I/O buffer only in full duplex mode
- fix function return code

6.1.6 - 2022-03-31
- Do not install/dist TOSE_loader.h if loader is disabled
- Fixed header doc in TOSE_setup.h: removed reference to TO-Protect

6.1.5 - 2022-03-22
- Fix pyutils driver config detection

6.1.3 - 2022-03-22
- Fix loader helper enabling

6.1.2 - 2022-03-16
- updated footprints values, and fixed table of contents

6.1.1 - 2022-03-15
- Changes for TO-Protect 2.0.0

6.0.2 - 2021-12-14
- doc: minor fix

6.0.1 - 2021-12-09
- TLS: improve documentation about HANDSHAKE ONLY mode
- add TO_ prefix to headers to avoid name conflicts with mbedOS

6.0.0 - 2021-12-03
TLS: fix invalid max_length when receiving data
- TLS: fix src/dst inversion in memmove
- TLS: fix bad handling of partial records
- TLS: add the default cipher aes-gcm for HANDSHAKE_ONLY mode
- TLS: deprecate TOSE_helper_tls_receive()
- Certs: fix TOSE_helper_get_certificate_x509_and_sign() output size
- TLS: fix regression when multiple handshake messages are contained inside a single record

5.14.1 - 2021-10-22
- TLS: fix buffer overflow in GCM/CCM modes when encrypting messages
- TLS: fix handshake failure with 0-length session_id in server_hello
- sphinx theme: fix configuration of html_static_path
- doc: set all doxygen group tags in lower case

5.14.0 - 2021-10-07
- TLS: rename 'ret2' to a name with a bit more meaning
- fix: regression with TLS HANDSHAKE_ONLY mode
- TLS SNI extension: add API to support SNI

5.13.0 - 2021-08-31
- install: adjust prefix of python's TOdir
- TLS: add python entry points to TLS API
- Fix typos spotted by Michel
- TLS: remove automatic mode selection in TLS 1.2
- TLS: add definitions for CMD_TLS_GET_MASTER_SECRET_DERIVED_KEYS
- TLS: remove stack allocated variable length array
- TLS: clean (a bit) the #ifdef mess
- TLS: improve (a bit) the documentation
- Add tls_send_record()
- TLS: calculate maximum plain text length to send once
- TLS: clean (a bit) the #ifdef mess
- Add TOSE_helper_tls_recv()
- helper_certs: add TOSE_helper_get_certificate_x509_and_sign()
- TLS_set_session: cache the current session
- TLS: move close_received state into global state
- Fixed buffer overflow described in TOSF-63

5.12.4 - 2021-07-12
- Fixed make install for autotools dist release, now run 'make full-dist-zip'
instead of 'make install' to generate dist-zip archive

5.12.3 - 2021-06-23
- use get_certificate_x509 init/update/final API through TOsh.py
- add get_certificate_x509 init/update/final API
- Doc improvment: Precise AES128 (not -CBC) is used for initial chaining vector, in admin commands
- Doc improvment: Add precisions to CRC byte ordering in Administration Commands

5.12.2 - 2021-04-08
- Broadcast loader helper: do not fail in case of retry and if the ongoing
migration has been already done

5.12.1 - 2021-04-08
- Add API to get Measured Boot SE measurement

5.12.0 - 2021-04-07
- Broadcast loader helper now supports TUP chaining, data migration tags, TUP

main.cpp
/**
 * THIS SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTIONs OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
 * DEALINGS IN THE SOFTWARE.
 *
 * Copyright 2018-2019 Trusted Objects
 *
 * @file main.cpp
 * @brief Trusted Objects' TLS Stack usage example
 */

#include "mbed.h"
#include "wifi.h"

#include "sample_https.h"

/* Private defines -----------------------------------------------------------*/
#define CONNECTION_TRIAL_MAX 5

enum
{
	TLS_CNX_FULL_HANDHAKE,
	TLS_CNX_RESUME_HANDHAKE,
};

/* Private macro -------------------------------------------------------------*/

#define TEST_AND_CHECK(action)                                    \
	{                                                         \
		int status;                                       \
		status = (action);                                \
		if (status != 0)                                  \
		{                                                 \
			LOG("DA_KO: %d, %d", __LINE__, status); \
			goto exit;                                    \
		}                                                 \
	}

/* Private prototypes --------------------------------------------------------*/

static int initialize_wifi(void);

static int connect_wifi(void);
static int disconnect_wifi(void);

static int connect_tls_server(int mode);
static int disconnect_tls_server(void);


/* Private variables ---------------------------------------------------------*/

static char http_request[] = "GET /index.html HTTP/1.1\r\n\r\n";

Serial pc(SERIAL_TX, SERIAL_RX);
DigitalIn button(USER_BUTTON);

static uint8_t RemoteIP[4];
static uint8_t MAC_Addr[6];
static uint8_t IP_Addr[4];

static int32_t Socket = -1;

Timer t_elapsed;

/**
 * @brief main function
 *
 * This main function will start a Command line interface
 * The main USB (emulating a RS232 line) is used
 *
 * @return
 * - 0: success
 * - 1: failure
 */
int main()
{
	pc.baud(115200);
	char *server_response;

	t_elapsed.start();

	LOG(" ");
	LOG("************************************");
	LOG("* Trusted Objects TLS Demonstrator *");
	LOG("************************************");

	// Enable writing to ROM while in main (mandatory when using STM32 HAL)
	ScopedRomWriteLock make_rom_writable;

	LOG("********** Initialize Secure Element ***************************");

	TEST_AND_CHECK(POC_init_se());
	TEST_AND_CHECK(POC_display_se_infos());

	LOG("********** Initialize and connect Wifi *************************");

	TEST_AND_CHECK(initialize_wifi());
	TEST_AND_CHECK(connect_wifi());

	LOG("********** Connect in TLS to remote server *********************");

	TEST_AND_CHECK(connect_tls_server(TLS_CNX_FULL_HANDHAKE));

	LOG("********** Send HTTP GET request an get `Hello world !` answer *");

	TEST_AND_CHECK(POC_http_post(MBED_CONF_APP_SERVER_HOSTNAME, http_request, strlen(http_request), &server_response));

	/* Here, HTTP server normally cut the connection already */
	goto end;

exit:
	disconnect_tls_server();

end:
	disconnect_wifi();

	POC_fini_se();

	LOG("Terminated ! Press RESET (black) button to restart");

	t_elapsed.stop();
	return 0;
}

/**
 * @brief Initializes the WiFi
 * @param none
 *
 * WiFi SSID & Password comes from mbed_app.json
 * and can can be overriden by pressing the User Button right after Reset
 *
 * @return
 * - 0: success
 * - -1: an error occured
 */
static int initialize_wifi(void)
{
	/*Initialize  WIFI module */
	if (WIFI_Init() == WIFI_STATUS_OK)
	{
		LOG("WIFI Module Initialized.");
	}
	else
	{
		LOG("ERROR: WIFI Module cannot be initialized.");
		return -1;
	}

	if (WIFI_GetMAC_Address(MAC_Addr) == WIFI_STATUS_OK)
	{
		LOG("es-wifi module MAC Address : %X:%X:%X:%X:%X:%X",
		    MAC_Addr[0],
		    MAC_Addr[1],
		    MAC_Addr[2],
		    MAC_Addr[3],
		    MAC_Addr[4],
		    MAC_Addr[5]);
	}
	else
	{
		LOG("ERROR: CANNOT get MAC address");
		return -1;
	}

	return 0;
}

/**
 * @brief Connects the WiFi
 * @param none
 *
 * WiFi SSID & Password comes from mbed_app.json
 * and can can be overriden by pressing the User Button right after Reset
 *
 * @return
 * - 0: success
 * - -1: an error occured
 */
static int connect_wifi(void)
{
	LOG("connecting to : %s", MBED_CONF_APP_WIFI_SSID);
	if (WIFI_Connect(MBED_CONF_APP_WIFI_SSID, MBED_CONF_APP_WIFI_PASSWORD, WIFI_ECN_WPA2_PSK) == WIFI_STATUS_OK)
	{
		LOG("es-wifi module connected");
		/* Get Device local Address */
		if (WIFI_GetIP_Address(IP_Addr) == WIFI_STATUS_OK)
		{
			LOG("es-wifi module got IP Address : %d.%d.%d.%d",
			    IP_Addr[0],
			    IP_Addr[1],
			    IP_Addr[2],
			    IP_Addr[3]);
		}
		else
		{
			LOG("ERROR: es-wifi module CANNOT get IP address");
			return -1;
		}
	}
	else
	{
		LOG("ERROR: es-wifi module CANNOT connect");
		return -1;
	}

	return 0;
}

/**
 * @brief Disconnects the WiFi
 * @param none
 *
 * @return
 * - 0: success
 * - -1: an error occured
 */
static int disconnect_wifi(void)
{
	if (WIFI_Disconnect() == WIFI_STATUS_OK)
	{
		LOG("es-wifi disconnected");
	}
	else
	{
		LOG("ERROR disconnecting WiFi ");
		return -1;
	}

	return 0;
}

/**
 * @brief Initiate a TLS connection with the remote server
 * @param mode  perform FULL or RESUME handshake  (TLS_CNX_FULL_HANDHAKE ...)
 *
 * Server HostName comes from mbed_app.json
 * and can can be overriden by pressing the User Button right after Reset
 *
 * @return
 * - 0: success
 * - -1: an error occured
 */


int connect_tls_server(int mode)
{
	uint16_t Trials = CONNECTION_TRIAL_MAX;

	/* Check that WiFi is connected first */
	if (WIFI_GetIP_Address(IP_Addr) != WIFI_STATUS_OK)
	{
		LOG("WiFi is NOT connected; You need to connect WiFi FIRST !!!");
		return -1;
	}

	/* Get Host Remote Address */
	LOG("Converting Server remote address from [%s]", MBED_CONF_APP_SERVER_HOSTNAME);
	if (WIFI_GetHostAddress((char *)MBED_CONF_APP_SERVER_HOSTNAME, RemoteIP) == WIFI_STATUS_OK)
	{
		LOG("Trying to connect to Server: %d.%d.%d.%d:%d ...",
		    RemoteIP[0],
		    RemoteIP[1],
		    RemoteIP[2],
		    RemoteIP[3],
		    MBED_CONF_APP_SERVER_PORT);
	}
	else
	{
		LOG("Exiting ... cannot get Server remote address from [%s]", MBED_CONF_APP_SERVER_HOSTNAME);
		return -1;
	}

	while (Trials)
	{
		if (WIFI_OpenClientConnection(0, WIFI_TCP_PROTOCOL, "TCP_CLIENT", RemoteIP, MBED_CONF_APP_SERVER_PORT, 0) == WIFI_STATUS_OK)
		{
			LOG("TCP Connection opened successfully.");
			Socket = 0;
			break;
		}
		Trials--;
	}
	if (!Trials)
	{
		LOG("ERROR: Cannot open Connection");
		return -1;
	}

	int errTO;
	int reset_tls;
	reset_tls = (mode == TLS_CNX_FULL_HANDHAKE) ? 1 : 0;
	// connect in TLS
	errTO = POC_connect_tls_server(static_cast<void *>(&Socket), reset_tls, (char *)MBED_CONF_APP_SERVER_HOSTNAME);
	if (errTO != 0)
	{
		LOG("TO_KO: %d, %d", __LINE__, errTO);
		return -1;
	}

	return 0;
}

/**
 * @brief Terminates a TLS connection with the remote server
 * @param none
 *
 * @return
 * - 0: success
 * - -1: an error occured
 */
static int disconnect_tls_server(void)
{
	if (Socket == -1)
	{
		LOG("You are NOT connected !");
		return -1;
	}

	POC_disconnect_tls_server(static_cast<void *>(&Socket));

	WIFI_CloseClientConnection(0);

	Socket = -1;

	return 0;
}

/**
 * @brief Display elapsed time
 *
 */

void POC_display_time(void)
{
	printf("[t=%f] ", t_elapsed.read());
}

